### PR TITLE
RATIS-1441.Add snapshot manually request and proto

### PR DESCRIPTION
**Pull request by codings-dan**
## What changes were proposed in this pull request?

subtask of Support snapshot command: Add snapshotManually request and proto

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1441

## How was this patch tested?
No need

### CODE REVIEW

**Comment by codings-dan** [comment]
@szetszwo My idea is to first complete this pr, add proto and request. Then manually trigger the snapshot on the server side, and finally add a command. Can you help review the code, thx!

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 1]
```diff
@@ -963,6 +963,15 @@ void finishTransferLeadership() {
     }
   }
 
+  public RaftClientReply snapshotManually(SnapshotManuallyRequest request) throws IOException {
+    return waitForReply(request, snapshotManuallyAsync(request));
+  }
+
+  public CompletableFuture<RaftClientReply> snapshotManuallyAsync(SnapshotManuallyRequest request) throws IOException {
+    //TODO(codings-dan) : implement snapshot manually in server
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:971]: @codings-dan , we should implement the feature (i.e. this TODO) in the server first and then change the API (this pull request).

**Review by szetszwo — COMMENTED** [review]
@codings-dan , Let's implement the feature first before changing the API.

BTW, let's remove "manually" from the names.   For example, "SnapshotManuallyRequest" can be simply called "SnapshotRequest".

I am also thinking about add a SnapshotManagementApi (which is similar to GroupManagementApi) so that we may add other operations later on.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java (hunk 1 above)
Comment (codings-dan) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:971]: Ok, I will try to implement this feature on the server side first

**Comment by codings-dan** [comment]
@szetszwo hi, I tried to implement takesnapshot on the server side, can you help review the code, thx!

**Comment by codings-dan** [comment]
will do

**Comment by codings-dan** [comment]
@szetszwo hi, since the implementation code on the server side depends on SnapshotRequest and some proto-related code, can we review and merge the proto-related commit first, and then complete the server-side modification, otherwise there will be many errors

**Comment by szetszwo** [comment]
@codings-dan , we may use the GroupManagementRequest as an example for the snapshot request:
- SnapshotRequest should be similar to GroupManagementRequest.
- add a takeSnapshotAsync(..) method in RaftServerProxy for taking a new snapshot; see RaftServerProxy.groupAddAsync(..).
- add new unit tests to call RaftServerProxy.takeSnapshotAsync(..) directly. 

The items above are the works for server change.

In the next JIRA, we will add RaftServerProxy.snapshotManagementAsync(SnapshotRequest) and also SnapshotRequestProto.  Then, we can change the unit tests to call from the client.

**Comment by codings-dan** [comment]
@szetszwo hi, now the server-side code has been implemented, and some unit tests have been added. Can we discuss proto and API change, it seems that it is not appropriate to put it under AdminApi, can we create a new API interface, such as snapshotManagementApi?

**Comment by szetszwo** [comment]
@codings-dan , yes, it is a good time to talk about the API.  Let's add a SnapshotManagementApi.  Currently, we only have the create(..) method.  In the future, we may consider adding other operations such as 
- remove
- list
- download
- upload

**Comment by szetszwo** [comment]
BTW, let's change the SimulatedRpc first and then change gRPC, Netty and Hadoop RPC later.

**Comment by codings-dan** [comment]
@szetszwo I change the code, added the code in SimulatedRpc and the raft.proto, PTAL, thx! In addition, the CI test error seems to have nothing to do with this code change.

File: ratis-common/src/main/java/org/apache/ratis/protocol/SnapshotManagementAsyncProtocol.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.protocol;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+public interface SnapshotManagementAsyncProtocol {
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/protocol/SnapshotManagementAsyncProtocol.java:23]: Similar to groupManagementAsync, let's create a snapshotManagementAsync  method in AdminAsynchronousProtocol instead adding a new protocol.
```
//AdminAsynchronousProtocol.java
  CompletableFuture<RaftClientReply> snapshotManagementAsync(SnapshotManagementRequest request);
```

File: ratis-common/src/main/java/org/apache/ratis/protocol/SnapshotManagementProtocol.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.protocol;
+
+import java.io.IOException;
+
+public interface SnapshotManagementProtocol {
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/protocol/SnapshotManagementProtocol.java:22]: Let's add a snapshotManagement method in AdminProtocol.
```
//AdminProtocol.java
  RaftClientReply snapshotManagement(SnapshotManagementRequest request) throws IOException;
```

File: ratis-proto/src/main/proto/Raft.proto [hunk 4]
```diff
@@ -413,6 +413,11 @@ message TransferLeadershipRequestProto {
   RaftPeerProto newLeader = 2;
 }
 
+// snapshot request
+message SnapshotRequestProto {
```
Comment (szetszwo) [inline ratis-proto/src/main/proto/Raft.proto:417]: Let's call it SnapshotManagementRequestProto.  We should also rename SnapshotRequest to SnapshotManagementRequest.

**Review by szetszwo — COMMENTED** [review]
@codings-dan , thanks for the update.  Instead of adding SnapshotManagement[Async]Protocol, let's add a new snapshotManagement[Async] method to Admin[Async]Protocol.

**Comment by codings-dan** [comment]
@szetszwo I have added snapshotmanagement related code to the admin related protocol, please take a look, thank youI have added snapshotmanagement related code to the admin related protocol, PTAL, thx!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by codings-dan** [comment]
@szetszwo In the next pull request, I will add gRPC, Netty and Hadoop related proto, and then I will add the corresponding shell command and add the corresponding unit test, what do you think?

**Comment by szetszwo** [comment]
@codings-dan , sure.  Let's add gRPC first.  We will add Netty and Hadoop later.  Thanks a lot!

**Pull request merged**